### PR TITLE
Autoroutines

### PR DESCRIPTION
**Pull request by LQ1234**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -70,4 +79,9 @@
     public static final int GRAYHILL_ENCODER_DISTANCE_PER_PULSE = 1; //TODO: Replace this value with a correct one 
 
     public static final int XBOX_A = 1;
+
+    // Vision Subsystem
+    public static final double TOLERANCE_DEG = 5.;
+    public static final double STABALIZE_SECS = .5;
+    public static final double TIMEOUT_SECS = 3;
```
Comment (prensing) [inline src/main/java/frc/robot/Constants.java:86]: These constants are not part of vision, they are part of the drivetrain system (used in TurnToHub). The names are not good because they are too generic. And "stabilize" is misspelled.

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -9,6 +9,7 @@
 import edu.wpi.first.wpilibj.DriverStation;
 import edu.wpi.first.wpilibj.GenericHID;
 import edu.wpi.first.wpilibj.XboxController;
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
```
Comment (prensing) [inline src/main/java/frc/robot/RobotContainer.java:12]: Why was this added when it is not used? While you are at it, please remove the other unused imports.

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -74,19 +75,44 @@ public double getAsDouble() {
     private class Turn implements DoubleSupplier {
         @Override
         public double getAsDouble() {
-            return -m_xbox.getRightX();
+            if (Robot.isReal())
+                return m_xbox.getRightX();
+            else
+                return m_xbox.getLeftX();
```
Comment (prensing) [inline src/main/java/frc/robot/RobotContainer.java:81]: First, you lost a minus sign on the controller value. And 2nd, no, leave the controller the same in the simulation.

File: src/main/java/frc/robot/commands/IntakeCommand.java [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 import edu.wpi.first.wpilibj2.command.CommandBase;
 import frc.robot.subsystems.Intake;
+import frc.robot.subsystems.Climber;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/IntakeCommand.java:5]: Why was this added?

File: src/main/java/frc/robot/commands/OneBallAuto.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import frc.robot.subsystems.Intake;
+import frc.robot.subsystems.Shooter;
+import frc.robot.subsystems.Vision;
+
+public class OneBallAuto extends SequentialCommandGroup implements AutoCommandInterface {
+
+    public OneBallAuto(Shooter shooter, Intake intake, DriveTrain driveTrain, Vision vision) {
+        addCommands(
+            new TurnTowardsHub(driveTrain, vision).withTimeout(Constants.TIMEOUT_SECS),
+            new SelectCommand(() -> new ShooterCommand(shooter, intake, vision, Shooter.calculateShooterSpeeds(vision.getDistance())))
```
Comment (prensing) [inline src/main/java/frc/robot/commands/OneBallAuto.java:18]: Why do you have a SelectCommand here (and all the others)? You are only running a fixed command.

File: src/main/java/frc/robot/commands/OneBallAuto.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import frc.robot.subsystems.DriveTrain;
+import frc.robot.subsystems.Intake;
+import frc.robot.subsystems.Shooter;
+import frc.robot.subsystems.Vision;
+
+public class OneBallAuto extends SequentialCommandGroup implements AutoCommandInterface {
+
+    public OneBallAuto(Shooter shooter, Intake intake, DriveTrain driveTrain, Vision vision) {
+        addCommands(
+            new TurnTowardsHub(driveTrain, vision).withTimeout(Constants.TIMEOUT_SECS),
```
Comment (prensing) [inline src/main/java/frc/robot/commands/OneBallAuto.java:17]: The timeout should be part of TurnTowardsHub, not added on everywhere. You are repeating the same code over and over.

File: src/main/java/frc/robot/commands/OneBallAuto.java (hunk 5 above)
Comment (prensing) [reply src/main/java/frc/robot/commands/OneBallAuto.java:18]: I don't like that the speeds are computed outside the command. This will make it hard to have any kind of backup plan if the vision is not working. I suggest we pass in Vision to the command constructor and then have it handle finding the distance and setting the speeds.

Comment (prensing) [reply src/main/java/frc/robot/commands/OneBallAuto.java:18]: Actually, it already has "Vision", so just move finding the speeds inside.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+import frc.robot.Constants;
+import frc.robot.FieldInformation;
+import frc.robot.subsystems.DriveTrain;
+import frc.robot.subsystems.Intake;
+import frc.robot.subsystems.Shooter;
+import frc.robot.subsystems.Vision;
+
+public class TwoBallAutoCurved extends SequentialCommandGroup implements AutoCommandInterface {
+    static final double DISTANCE_BACK = 1.4;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:27]: Not used.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    Trajectory trajectory;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:28]: Wrong variable name.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            new PIDController(Constants.kPDriveVel, 0, 0),
+            driveTrain::tankDriveVolts,
+            driveTrain
+        );
+
+        addCommands(
+            new ParallelDeadlineGroup(
+                new SequentialCommandGroup(
+                    ramseteCommand.andThen(() -> driveTrain.tankDriveVolts(0, 0)),
+                    new WaitCommand(0.5) // for the ball to enter the hopper
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:71]: Don't put in arbitrary waits.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                ), 
+                new IntakeCommand(intake, Constants.INTAKE_SHOOTING_SPEED)
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:73]: This is not "shooting speed", it should be "intake speed". Might end up being the same, but should not be coded with the same variable.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
+package frc.robot.commands;
+
+import java.util.List;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.controller.RamseteController;
+import edu.wpi.first.math.controller.SimpleMotorFeedforward;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:9]: This import is not used. Please try to clean up all warnings.

File: src/main/java/frc/robot/commands/TwoBallAutoStraight.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            new TurnTowardsHub(driveTrain, vision).withTimeout(Constants.TIMEOUT_SECS),
+            new SelectCommand(() -> new ShooterCommand(shooter, intake, vision, Shooter.calculateShooterSpeeds(vision.getDistance())))
+        );
+    }
+
+    @Override
+    public Pose2d getInitialPose() {
+        return FieldInformation.lowerBlueStart;
+    }
+}
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoStraight.java:88]: Why do you not plot the trajectory?

File: src/main/java/frc/robot/commands/TwoBallAutoStraight.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
+package frc.robot.commands;
+
+import java.util.List;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.controller.RamseteController;
+import edu.wpi.first.math.controller.SimpleMotorFeedforward;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoStraight.java:9]: unused.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        addCommands(
+            new ParallelDeadlineGroup(
+                new SequentialCommandGroup(
+                    ramseteCommand.andThen(() -> driveTrain.tankDriveVolts(0, 0)),
+                    new WaitCommand(0.5) // for the ball to enter the hopper
+                ), 
+                new IntakeCommand(intake, Constants.INTAKE_SHOOTING_SPEED)
+            ),
+            new TurnTowardsHub(driveTrain, vision).withTimeout(Constants.TIMEOUT_SECS),
+            new SelectCommand(() -> new ShooterCommand(shooter, intake, vision, Shooter.calculateShooterSpeeds(vision.getDistance())))
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:76]: Again I don't see a reason for SelectCommand

File: src/main/java/frc/robot/commands/TwoBallAutoStraight.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            new ParallelDeadlineGroup(
+                new SequentialCommandGroup(
+                    ramseteCommand.andThen(() -> driveTrain.tankDriveVolts(0, 0)),
+                    new WaitCommand(0.5) // for the ball to enter the hopper
+                ), 
+                new IntakeCommand(intake, Constants.INTAKE_SHOOTING_SPEED)
+            ),
+            new TurnTowardsHub(driveTrain, vision).withTimeout(Constants.TIMEOUT_SECS),
+            new SelectCommand(() -> new ShooterCommand(shooter, intake, vision, Shooter.calculateShooterSpeeds(vision.getDistance())))
+        );
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TwoBallAutoStraight.java:81]: See Curved for the same comments.

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 16]
```diff
@@ -19,25 +22,50 @@
...
-    //CANSparkMax for the hopper
-    CANSparkMax m_chuteMotor;
-    //WPI_TalonFX for the shooter
-    WPI_TalonFX m_topShooterMotor, m_bottomShooterMotor;
-    //Limit Switch for Intake
-    DigitalInput m_limitSwitch1, m_limitSwitch2;
-
+    CANSparkMax m_chuteMotor; //CANSparkMax for the hopper
+    WPI_TalonFX m_topShooterMotor, m_bottomShooterMotor; //WPI_TalonFX for the shooter
+    DigitalInput m_limitSwitch1, m_limitSwitch2; //Limit Swit
```
Comment (prensing) [inline src/main/java/frc/robot/subsystems/Shooter.java:27]: Please put the comments back as they are in "main".

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 17]
```diff
@@ -19,25 +22,50 @@
...
+                bottom * (1 - ratio) + other.bottom * ratio,
+                chute * (1 - ratio) + other.chute * ratio
+            );
+        }
+    }
+
+    public static ShooterSpeeds calculateShooterSpeeds(double distance){
+        Map.Entry<Double, ShooterSpeeds> before = shooterSpeeds.floorEntry(distance);
+        Map.Entry<Double, ShooterSpeeds> after = shooterSpeeds.ceilingEntry(distance);
+        if (before == null && after == null) throw new Error("Shooter Speeds empty");
```
Comment (prensing) [inline src/main/java/frc/robot/subsystems/Shooter.java:64]: No. You **cannot** throw an error. This will crash the whole robot code.

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 18]
```diff
@@ -1,9 +1,17 @@
...
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.util.Units;
 import edu.wpi.first.wpilibj.Relay;
+import edu.wpi.first.wpilibj.Timer;
 import edu.wpi.first.wpilibj.Relay.Value;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.CommandBase;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.FieldInformation;
+import frc.robot.Robot;
```
Comment (prensing) [inline src/main/java/frc/robot/subsystems/Vision.java:14]: Not used. Please try to clean up *all* warnings.

File: src/main/java/frc/robot/commands/TurnTowardsHub.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+            m_withinTolerance = Math.abs(errorDeg) < m_toleranceDeg;
+            if(!m_withinTolerance) m_lastOutsideToleranceTime = Timer.getFPGATimestamp();
+            m_driveTrain.drive(0, -errorDeg * kP, false);
+        }
+
+        // Called once the command ends or is interrupted.
+        @Override
+        public void end(boolean interrupted) {
+            m_driveTrain.drive(0, 0, false);
+            m_vision.setMode(Vision.VisionMode.INTAKE);
```
Comment (prensing) [inline src/main/java/frc/robot/commands/TurnTowardsHub.java:41]: You can't do this here. This will turn on the finder and the ShooterCommand won't be able to get the distance.

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 20]
```diff
@@ -61,26 +92,46 @@ public VisionMode getMode() {
...
-        double[] visionData = SmartDashboard.getNumberArray("vision/target_info", EMPTY_TARGET_INFO);
-        return Math.toDegrees(visionData[4]);
+        return Math.toDegrees(m_lastSuccessfulResult[4]);
     }
 
     public void setLedRing (boolean on) {
         m_spike.set(on ? Value.kReverse : Value.kOff);
     }
+
+    public class AwaitVisionResult extends CommandBase {
```
Comment (prensing) [inline src/main/java/frc/robot/subsystems/Vision.java:115]: This should be a separate command in with all the other commands. Also you can't reset the mode at the end, otherwise the Shooter will not get a distance.

**Review by prensing — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 21]
```diff
@@ -19,25 +22,50 @@
...
+    public static class ShooterSpeeds {
+        double top, bottom, chute;
+        public ShooterSpeeds(double top, double bottom, double chute) {
+            this.top = top;
+            this.bottom = bottom;
+            this.chute = chute;
+        }
+        public ShooterSpeeds interpolate(ShooterSpeeds other, double ratio) {
+            return new ShooterSpeeds(
+                top * (1 - ratio) + other.top * ratio,
```
Comment (prensing) [inline src/main/java/frc/robot/subsystems/Shooter.java:54]: A cleaner and _very slightly_ faster expression would be:
   top + (other.top - top) * ratio
To me, this is much more understandable.

File: src/main/java/frc/robot/commands/TurnTowardsHub.java (hunk 19 above)
Comment (LQ1234) [reply src/main/java/frc/robot/commands/TurnTowardsHub.java:41]: I think the shooter command can use the last vision result (which is cashed)

File: src/main/java/frc/robot/subsystems/Vision.java (hunk 20 above)
Comment (LQ1234) [reply src/main/java/frc/robot/subsystems/Vision.java:115]: It is only reset if it is interrupted (because in a sequential command group the command that's running is responsible for cleaning up)

File: src/main/java/frc/robot/commands/AwaitVisionResult.java [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    @Override
+    public boolean isFinished() {
+        return m_vision.getStatus();
+    }
+    @Override
+    public void execute() {
+    }
+    @Override
+    public void end(boolean interrupted) {
+        if(interrupted) m_vision.setMode(Vision.DEFAULT_MODE);
```
Comment (coachFitz0807) [inline src/main/java/frc/robot/commands/AwaitVisionResult.java:27]: Since we didn't define a required subsystem for this command, it will neve be interrupted.

File: src/main/java/frc/robot/commands/TurnTowardsHub.java (hunk 19 above)
Comment (coachFitz0807) [reply src/main/java/frc/robot/commands/TurnTowardsHub.java:41]: We shouldn't change the VIsionMode here. We should not change the VisionMode until after we shoot, so this call should go in the end() method of the ShooterCommand.

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+            driveTrain::tankDriveVolts,
+            driveTrain
+        );
+
+        addCommands(
+            new ParallelDeadlineGroup(
+                ramseteCommand.andThen(() -> driveTrain.tankDriveVolts(0, 0)),
+                new IntakeCommand(intake, Constants.INTAKE_SPEED)
+            ),
+            new TurnTowardsHub(driveTrain, vision).withTimeout(Constants.TURN_TIMEOUT_SECS),
```
Comment (coachFitz0807) [inline src/main/java/frc/robot/commands/TwoBallAutoCurved.java:68]: The camera will be oriented with a large vertical FOV width and a narrow horizontal FOX width.
There is no guarantee that the robot will be able to see the target when it picks up the middle ball.
We should include a ramsete to turn back toward the hub before switching to Vision.

File: src/main/java/frc/robot/commands/TwoBallAutoStraight.java [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+
+        TrajectoryConfig config =
+            new TrajectoryConfig(Constants.kMaxSpeed,
+                                Constants.kMaxAcceleration)
+                .setKinematics(Constants.kDriveKinematics)
+                .addConstraint(autoVoltageConstraint)
+                .setReversed(true);
+
+        Pose2d initialPose = getInitialPose();
+        Pose2d finalPose = new Pose2d(
```
Comment (coachFitz0807) [inline src/main/java/frc/robot/commands/TwoBallAutoStraight.java:43]: If you use the FieldInformation to get the location of the ball, we can use the same command for all TwoBallAuto commands.

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 25]
```diff
@@ -18,26 +21,53 @@
...
-
-    //CANSparkMax for the hopper
+    // CANSparkMax for the hopper
     CANSparkMax m_chuteMotor;
-    //WPI_TalonFX for the shooter
+    // WPI_TalonFX for the shooter
     WPI_TalonFX m_topShooterMotor, m_bottomShooterMotor;
-    //Limit Switch for Intake
+    // Limit Switch for Intake
     DigitalInput m_limitSwitch1, m_limitSwitch2;
```
Comment (coachFitz0807) [inline src/main/java/frc/robot/subsystems/Shooter.java:29]: Add TODO comment to change these to Color Sensors.

**Review by coachFitz0807 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/TurnTowardsHub.java (hunk 19 above)
Comment (prensing) [reply src/main/java/frc/robot/commands/TurnTowardsHub.java:41]: I would prefer that ShooterCommand be generic and only deal with shooting balls. But definitely, the resetting of the Vision mode should be done as part of the cleanup for the meta-command which turns and shoots.

**Review by prensing — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/TwoBallAutoCurved.java (hunk 23 above)
Comment (prensing) [reply src/main/java/frc/robot/commands/TwoBallAutoCurved.java:68]: John, if you run the current auto in simulation, the robot ends pointing pretty accurately to the hub, so I think until we test, this is OK.

File: src/main/java/frc/robot/commands/AwaitVisionResult.java (hunk 22 above)
Comment (LQ1234) [reply src/main/java/frc/robot/commands/AwaitVisionResult.java:27]: In the autos the command is wrapped with withTimeout so that it will be interrupted if it takes too long

File: src/main/java/frc/robot/commands/TwoBallAutoStraight.java (hunk 24 above)
Comment (LQ1234) [reply src/main/java/frc/robot/commands/TwoBallAutoStraight.java:43]: I'm not sure what you mean?

File: src/main/java/frc/robot/commands/AwaitVisionResult.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (LQ1234) [reply src/main/java/frc/robot/commands/AwaitVisionResult.java:27]: So should I override the end method of the SequentialCommandGroups so that they also change the mode?

**Review by prensing — APPROVED** [review]
I am good with this now. Coach Fitz needs to also approve.

**Review by coachFitz0807 — APPROVED** [review]
